File: fix_it_fred_sdk.js
/**
 * Fix It Fred SDK - JavaScript/TypeScript
 * Universal AI Assistant Integration for Web Applications
 */

class FixItFredSDK {
    /**
     * Initialize Fix It Fred SDK
     * @param {string} baseUrl - Base URL of Fix It Fred service
     * @param {string} appId - Your application ID (optional)
     * @param {string} apiKey - Your API key (optional)
     */
    constructor(baseUrl = 'http://localhost:8005', appId = null, apiKey = null) {
        this.baseUrl = baseUrl.replace(/\/$/, '');
        this.appId = appId;
        this.apiKey = apiKey;
        
        this.defaultHeaders = {
            'Content-Type': 'application/json',
            'User-Agent': 'FixItFred-SDK-JS/1.0'
        };
        
        if (this.apiKey) {
            this.defaultHeaders['Authorization'] = `Bearer ${this.apiKey}`;
        }
    }
    
    /**
     * Send a chat message to Fix It Fred
     * @param {Object} options - Chat options
     * @param {string} options.message - The message to send
     * @param {string} options.provider - AI provider ('anthropic', 'openai', 'xai', 'google', 'ollama')
     * @param {string} options.context - Additional context
     * @param {string} options.conversationId - Conversation ID for tracking
     * @param {string} options.model - Specific model to use
     * @param {number} options.temperature - Response randomness (0.0-1.0)
     * @param {number} options.maxTokens - Maximum response length
     * @returns {Promise<Object>} Response object
     */
    async chat({
        message,
        provider = 'anthropic',
        context = null,
        conversationId = null,
        model = null,
        temperature = 0.7,
        maxTokens = 2000
    }) {
        const endpoint = this.appId 
            ? `${this.baseUrl}/api/apps/${this.appId}/chat`
            : `${this.baseUrl}/api/universal/chat`;
            
        const payload = {
            message,
            provider,
            context,
            conversation_id: conversationId,
            model,
            temperature,
            max_tokens: maxTokens
        };
        
        // Remove null values
        Object.keys(payload).forEach(key => {
            if (payload[key] === null || payload[key] === undefined) {
                delete payload[key];
            }
        });
        
        try {
            const response = await fetch(endpoint, {
                method: 'POST',
                headers: this.defaultHeaders,
                body: JSON.stringify(payload)
            });
            
            if (!response.ok) {
                throw new Error(`HTTP ${response.status}: ${response.statusText}`);
            }
            
            return await response.json();
        } catch (error) {
            return {
                success: false,
                error: error.message,
                response: `❌ Error connecting to Fix It Fred: ${error.message}`
            };
        }
    }
    
    /**
     * Register your application with Fix It Fred
     * @param {Object} config - Registration configuration
     * @returns {Promise<Object>} Registration result
     */
    async registerApp({
        appName,
        webhookUrl = null,
        allowedProviders = ['anthropic', 'openai'],
        defaultProvider = 'anthropic',
        customInstructions = null
    }) {
        if (!this.appId) {
            this.appId = this.generateUUID();
        }
        
        const payload = {
            app_id: this.appId,
            app_name: appName,
            webhook_url: webhookUrl,
            allowed_providers: allowedProviders,
            default_provider: defaultProvider,
            custom_instructions: customInstructions
        };
        
        try {
            const response = await fetch(`${this.baseUrl}/api/integrations/register`, {
                method: 'POST',
                headers: this.defaultHeaders,
                body: JSON.stringify(payload)
            });
            
            const result = await response.json();
            
            if (result.success) {
                this.appId = result.app_id;
            }
            
            return result;
        } catch (error) {
            return {
                success: false,
                error: error.message
            };
        }
    }
    
    /**
     * Get all available AI providers
     * @returns {Promise<Object>} Providers information
     */
    async getProviders() {
        try {
            const response = await fetch(`${this.baseUrl}/api/providers`, {
                headers: this.defaultHeaders
            });
            return await response.json();
        } catch (error) {
            return {
                success: false,
                error: error.message
            };
        }
    }
    
    /**
     * Get conversation history
     * @param {string} conversationId - Conversation ID
     * @returns {Promise<Object>} Conversation data
     */
    async getConversation(conversationId) {
        try {
            const response = await fetch(`${this.baseUrl}/api/conversations/${conversationId}`, {
                headers: this.defaultHeaders
            });
            return await response.json();
        } catch (error) {
            return {
                success: false,
                error: error.message
            };
        }
    }
    
    /**
     * Clear conversation history
     * @param {string} conversationId - Conversation ID
     * @returns {Promise<Object>} Success status
     */
    async clearConversation(conversationId) {
        try {
            const response = await fetch(`${this.baseUrl}/api/conversations/${conversationId}`, {
                method: 'DELETE',
                headers: this.defaultHeaders
            });
            return await response.json();
        } catch (error) {
            return {
                success: false,
                error: error.message
            };
        }
    }
    
    /**
     * Check service health
     * @returns {Promise<Object>} Health status
     */
    async healthCheck() {
        try {
            const response = await fetch(`${this.baseUrl}/health`, {
                headers: this.defaultHeaders
            });
            return await response.json();
        } catch (error) {
            return {
                status: 'unhealthy',
                error: error.message
            };
        }
    }
    
    /**
     * Get service statistics
     * @returns {Promise<Object>} Usage statistics
     */
    async getStats() {
        try {
            const response = await fetch(`${this.baseUrl}/api/stats`, {
                headers: this.defaultHeaders
            });
            return await response.json();
        } catch (error) {
            return {
                success: false,
                error: error.message
            };
        }
    }
    
    /**
     * Generate a simple UUID
     * @returns {string} UUID
     */
    generateUUID() {
        return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
            const r = Math.random() * 16 | 0;
            const v = c == 'x' ? r : (r & 0x3 | 0x8);
            return v.toString(16);
        });
    }
}

/**
 * Simple chat interface for Fix It Fred
 */
class FixItFredChat {
    /**
     * Initialize simple chat interface
     * @param {string} provider - Default AI provider
     * @param {string} baseUrl - Base URL of service
     */
    constructor(provider = 'anthropic', baseUrl = 'http://localhost:8005') {
        this.sdk = new FixItFredSDK(baseUrl);
        this.provider = provider;
        this.conversationId = this.sdk.generateUUID();
    }
    
    /**
     * Ask Fix It Fred a question
     * @param {string} message - Question or message
     * @param {string} context - Additional context
     * @returns {Promise<string>} Response text
     */
    async ask(message, context = null) {
        const result = await this.sdk.chat({
            message,
            context,
            provider: this.provider,
            conversationId: this.conversationId
        });
        
        if (result.success) {
            return result.response || 'No response received';
        } else {
            return `Error: ${result.error || 'Unknown error'}`;
        }
    }
    
    /**
     * Switch AI provider
     * @param {string} provider - New provider to use
     */
    switchProvider(provider) {
        this.provider = provider;
    }
    
    /**
     * Start a new conversation
     */
    newConversation() {
        this.conversationId = this.sdk.generateUUID();
    }
}

/**
 * React Hook for Fix It Fred integration
 */
function useFixItFred(provider = 'anthropic', baseUrl = 'http://localhost:8005') {
    if (typeof React === 'undefined') {
        throw new Error('React is required for useFixItFred hook');
    }
    
    const [fred] = React.useState(() => new FixItFredChat(provider, baseUrl));
    const [loading, setLoading] = React.useState(false);
    const [error, setError] = React.useState(null);
    
    const ask = React.useCallback(async (message, context = null) => {
        setLoading(true);
        setError(null);
        
        try {
            const response = await fred.ask(message, context);
            setLoading(false);
            return response;
        } catch (err) {
            setError(err.message);
            setLoading(false);
            return `Error: ${err.message}`;
        }
    }, [fred]);
    
    const switchProvider = React.useCallback((newProvider) => {
        fred.switchProvider(newProvider);
    }, [fred]);
    
    const newConversation = React.useCallback(() => {
        fred.newConversation();
        setError(null);
    }, [fred]);
    
    return {
        ask,
        switchProvider,
        newConversation,
        loading,
        error,
        provider: fred.provider
    };
}

/**
 * Web Component for Fix It Fred Chat
 */
class FixItFredChatComponent extends HTMLElement {
    constructor() {
        super();
        this.attachShadow({ mode: 'open' });
        this.fred = new FixItFredChat();
        this.render();
        this.setupEventListeners();
    }
    
    render() {
        this.shadowRoot.innerHTML = `
            <style>
                :host {
                    display: block;
                    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', sans-serif;
                }
                .chat-container {
                    border: 1px solid #ddd;
                    border-radius: 8px;
                    height: 400px;
                    display: flex;
                    flex-direction: column;
                }
                .messages {
                    flex: 1;
                    overflow-y: auto;
                    padding: 16px;
                }
                .message {
                    margin-bottom: 12px;
                }
                .user-message {
                    text-align: right;
                }
                .ai-message {
                    text-align: left;
                }
                .message-bubble {
                    display: inline-block;
                    padding: 8px 12px;
                    border-radius: 16px;
                    max-width: 70%;
                }
                .user-bubble {
                    background: #007bff;
                    color: white;
                }
                .ai-bubble {
                    background: #f8f9fa;
                    border: 1px solid #dee2e6;
                }
                .input-area {
                    padding: 16px;
                    border-top: 1px solid #ddd;
                    display: flex;
                    gap: 8px;
                }
                .message-input {
                    flex: 1;
                    padding: 8px 12px;
                    border: 1px solid #ddd;
                    border-radius: 4px;
                }
                .send-button {
                    padding: 8px 16px;
                    background: #007bff;
                    color: white;
                    border: none;
                    border-radius: 4px;
                    cursor: pointer;
                }
                .send-button:hover {
                    background: #0056b3;
                }
                .send-button:disabled {
                    background: #6c757d;
                    cursor: not-allowed;
                }
            </style>
            <div class="chat-container">
                <div class="messages" id="messages">
                    <div class="message ai-message">
                        <div class="message-bubble ai-bubble">
                            Hello! I'm Fix It Fred. How can I help you today?
                        </div>
                    </div>
                </div>
                <div class="input-area">
                    <input type="text" class="message-input" id="messageInput" 
                           placeholder="Type your message..." />
                    <button class="send-button" id="sendButton">Send</button>
                </div>
            </div>
        `;
    }
    
    setupEventListeners() {
        const messageInput = this.shadowRoot.getElementById('messageInput');
        const sendButton = this.shadowRoot.getElementById('sendButton');
        const messages = this.shadowRoot.getElementById('messages');
        
        const sendMessage = async () => {
            const message = messageInput.value.trim();
            if (!message) return;
            
            // Add user message
            this.addMessage(message, 'user');
            messageInput.value = '';
            sendButton.disabled = true;
            
            // Add loading message
            const loadingId = this.addMessage('Thinking...', 'ai', true);
            
            try {
                const response = await this.fred.ask(message);
                this.removeMessage(loadingId);
                this.addMessage(response, 'ai');
            } catch (error) {
                this.removeMessage(loadingId);
                this.addMessage(`Error: ${error.message}`, 'ai');
            }
            
            sendButton.disabled = false;
            messageInput.focus();
        };
        
        sendButton.addEventListener('click', sendMessage);
        messageInput.addEventListener('keypress', (e) => {
            if (e.key === 'Enter') {
                sendMessage();
            }
        });
    }
    
    addMessage(text, type, isLoading = false) {
        const messages = this.shadowRoot.getElementById('messages');
        const messageId = `msg-${Date.now()}`;
        
        const messageDiv = document.createElement('div');
        messageDiv.className = `message ${type}-message`;
        messageDiv.id = messageId;
        
        const bubbleClass = type === 'user' ? 'user-bubble' : 'ai-bubble';
        const icon = isLoading ? '🤔' : (type === 'user' ? '👤' : '🤖');
        
        messageDiv.innerHTML = `
            <div class="message-bubble ${bubbleClass}">
                ${icon} ${text}
            </div>
        `;
        
        messages.appendChild(messageDiv);
        messages.scrollTop = messages.scrollHeight;
        
        return messageId;
    }
    
    removeMessage(messageId) {
        const message = this.shadowRoot.getElementById(messageId);
        if (message) {
            message.remove();
        }
    }
}

// Register web component
if (typeof customElements !== 'undefined') {
    customElements.define('fix-it-fred-chat', FixItFredChatComponent);
}

// Export for different module systems
if (typeof module !== 'undefined' && module.exports) {
    // Node.js
    module.exports = { FixItFredSDK, FixItFredChat };
} else if (typeof define === 'function' && define.amd) {
    // AMD
    define([], function() {
        return { FixItFredSDK, FixItFredChat };
    });
} else if (typeof window !== 'undefined') {
    // Browser global
    window.FixItFredSDK = FixItFredSDK;
    window.FixItFredChat = FixItFredChat;
    window.useFixItFred = useFixItFred;
}

// Usage examples:

/*
// Basic usage
const fred = new FixItFredChat();
const response = await fred.ask("What's the weather like?");
console.log(response);

// Advanced SDK usage
const sdk = new FixItFredSDK();
await sdk.registerApp({
    appName: "My Web App",
    customInstructions: "Provide helpful responses for web application users"
});

const result = await sdk.chat({
    message: "Help me with JavaScript",
    provider: "anthropic",
    context: "User is learning web development"
});

// React Hook usage
function MyComponent() {
    const { ask, loading, error } = useFixItFred();
    
    const handleSubmit = async (message) => {
        const response = await ask(message);
        console.log(response);
    };
    
    return (
        <div>
            {loading && <p>Loading...</p>}
            {error && <p>Error: {error}</p>}
            <button onClick={() => handleSubmit("Hello!")}>
                Ask Fix It Fred
            </button>
        </div>
    );
}

// Web Component usage
// Just add <fix-it-fred-chat></fix-it-fred-chat> to your HTML
*/

File: fix_it_fred_sdk.py
#!/usr/bin/env python3
"""
Fix It Fred SDK - Universal AI Assistant Integration
===================================================
Easy integration library for any Python application
"""

import requests
import json
from typing import Optional, Dict, Any, List
from datetime import datetime
import uuid

class FixItFredSDK:
    """
    Python SDK for integrating Fix It Fred Universal AI Assistant
    """
    
    def __init__(self, base_url: str = "http://localhost:8005", app_id: str = None, api_key: str = None):
        """
        Initialize Fix It Fred SDK
        
        Args:
            base_url: Base URL of Fix It Fred service
            app_id: Your application ID (optional, for app-specific features)
            api_key: Your API key (optional, for authenticated requests)
        """
        self.base_url = base_url.rstrip('/')
        self.app_id = app_id
        self.api_key = api_key
        self.session = requests.Session()
        
        # Set default headers
        self.session.headers.update({
            'Content-Type': 'application/json',
            'User-Agent': 'FixItFred-SDK/1.0'
        })
        
        if self.api_key:
            self.session.headers.update({
                'Authorization': f'Bearer {self.api_key}'
            })
    
    def chat(self, 
             message: str, 
             provider: str = "anthropic",
             context: str = None,
             conversation_id: str = None,
             model: str = None,
             temperature: float = 0.7,
             max_tokens: int = 2000) -> Dict[str, Any]:
        """
        Send a chat message to Fix It Fred
        
        Args:
            message: The message to send
            provider: AI provider to use ('anthropic', 'openai', 'xai', 'google', 'ollama')
            context: Additional context for the conversation
            conversation_id: ID for conversation tracking
            model: Specific model to use (optional)
            temperature: Response randomness (0.0-1.0)
            max_tokens: Maximum response length
            
        Returns:
            Dict containing the response and metadata
        """
        endpoint = f"{self.base_url}/api/universal/chat"
        
        # Use app-specific endpoint if app_id is configured
        if self.app_id:
            endpoint = f"{self.base_url}/api/apps/{self.app_id}/chat"
        
        payload = {
            "message": message,
            "provider": provider,
            "context": context,
            "conversation_id": conversation_id,
            "model": model,
            "temperature": temperature,
            "max_tokens": max_tokens
        }
        
        # Remove None values
        payload = {k: v for k, v in payload.items() if v is not None}
        
        try:
            response = self.session.post(endpoint, json=payload)
            response.raise_for_status()
            return response.json()
        except requests.RequestException as e:
            return {
                "success": False,
                "error": str(e),
                "response": f"❌ Error connecting to Fix It Fred: {str(e)}"
            }
    
    def register_app(self, 
                     app_name: str,
                     webhook_url: str = None,
                     allowed_providers: List[str] = None,
                     default_provider: str = "anthropic",
                     custom_instructions: str = None) -> Dict[str, Any]:
        """
        Register your application with Fix It Fred
        
        Args:
            app_name: Name of your application
            webhook_url: URL for webhook notifications (optional)
            allowed_providers: List of AI providers your app can use
            default_provider: Default AI provider for your app
            custom_instructions: Custom instructions for AI interactions
            
        Returns:
            Dict containing registration details including app_id
        """
        if not self.app_id:
            self.app_id = str(uuid.uuid4())
        
        payload = {
            "app_id": self.app_id,
            "app_name": app_name,
            "webhook_url": webhook_url,
            "allowed_providers": allowed_providers or ["anthropic", "openai"],
            "default_provider": default_provider,
            "custom_instructions": custom_instructions
        }
        
        try:
            response = self.session.post(f"{self.base_url}/api/integrations/register", json=payload)
            response.raise_for_status()
            result = response.json()
            
            # Store app_id for future requests
            if result.get("success"):
                self.app_id = result.get("app_id")
            
            return result
        except requests.RequestException as e:
            return {
                "success": False,
                "error": str(e)
            }
    
    def get_providers(self) -> Dict[str, Any]:
        """
        Get all available AI providers
        
        Returns:
            Dict containing available providers and their details
        """
        try:
            response = self.session.get(f"{self.base_url}/api/providers")
            response.raise_for_status()
            return response.json()
        except requests.RequestException as e:
            return {
                "success": False,
                "error": str(e)
            }
    
    def get_conversation(self, conversation_id: str) -> Dict[str, Any]:
        """
        Get conversation history
        
        Args:
            conversation_id: ID of the conversation
            
        Returns:
            Dict containing conversation messages
        """
        try:
            response = self.session.get(f"{self.base_url}/api/conversations/{conversation_id}")
            response.raise_for_status()
            return response.json()
        except requests.RequestException as e:
            return {
                "success": False,
                "error": str(e)
            }
    
    def clear_conversation(self, conversation_id: str) -> Dict[str, Any]:
        """
        Clear conversation history
        
        Args:
            conversation_id: ID of the conversation to clear
            
        Returns:
            Dict containing success status
        """
        try:
            response = self.session.delete(f"{self.base_url}/api/conversations/{conversation_id}")
            response.raise_for_status()
            return response.json()
        except requests.RequestException as e:
            return {
                "success": False,
                "error": str(e)
            }
    
    def health_check(self) -> Dict[str, Any]:
        """
        Check if Fix It Fred service is healthy
        
        Returns:
            Dict containing health status
        """
        try:
            response = self.session.get(f"{self.base_url}/health")
            response.raise_for_status()
            return response.json()
        except requests.RequestException as e:
            return {
                "status": "unhealthy",
                "error": str(e)
            }
    
    def get_stats(self) -> Dict[str, Any]:
        """
        Get service statistics
        
        Returns:
            Dict containing usage statistics
        """
        try:
            response = self.session.get(f"{self.base_url}/api/stats")
            response.raise_for_status()
            return response.json()
        except requests.RequestException as e:
            return {
                "success": False,
                "error": str(e)
            }

# Convenience class for simple chat interactions
class FixItFredChat:
    """
    Simple chat interface for Fix It Fred
    """
    
    def __init__(self, provider: str = "anthropic", base_url: str = "http://localhost:8005"):
        self.sdk = FixItFredSDK(base_url=base_url)
        self.provider = provider
        self.conversation_id = str(uuid.uuid4())
    
    def ask(self, message: str, context: str = None) -> str:
        """
        Ask Fix It Fred a question and get the response text
        
        Args:
            message: Question or message
            context: Additional context
            
        Returns:
            Response text from AI
        """
        result = self.sdk.chat(
            message=message,
            context=context,
            provider=self.provider,
            conversation_id=self.conversation_id
        )
        
        if result.get("success"):
            return result.get("response", "No response received")
        else:
            return f"Error: {result.get('error', 'Unknown error')}"
    
    def switch_provider(self, provider: str):
        """
        Switch AI provider
        
        Args:
            provider: New provider to use
        """
        self.provider = provider
    
    def new_conversation(self):
        """
        Start a new conversation (clear history)
        """
        self.conversation_id = str(uuid.uuid4())

# Integration examples for different use cases
class WebAppIntegration:
    """
    Example integration for web applications
    """
    
    def __init__(self, app_name: str):
        self.fred = FixItFredSDK()
        self.app_name = app_name
        
        # Register the app
        registration = self.fred.register_app(
            app_name=app_name,
            custom_instructions="You are integrated into a web application. Provide helpful, concise responses."
        )
        
        if registration.get("success"):
            print(f"✅ {app_name} registered with Fix It Fred")
        else:
            print(f"❌ Failed to register {app_name}")
    
    def handle_user_query(self, user_message: str, user_context: str = None) -> Dict[str, Any]:
        """
        Handle a user query from the web app
        """
        return self.fred.chat(
            message=user_message,
            context=user_context,
            provider="anthropic"  # Use most reliable provider for web apps
        )

class ChatbotIntegration:
    """
    Example integration for chatbots
    """
    
    def __init__(self, bot_name: str, platform: str = "discord"):
        self.fred = FixItFredSDK()
        self.bot_name = bot_name
        self.platform = platform
        
        # Register with platform-specific instructions
        platform_instructions = {
            "discord": "You are integrated into a Discord bot. Keep responses under 2000 characters and use Discord markdown when helpful.",
            "slack": "You are integrated into a Slack bot. Use Slack formatting and keep responses concise for workplace communication.",
            "telegram": "You are integrated into a Telegram bot. Provide helpful responses suitable for mobile messaging."
        }
        
        self.fred.register_app(
            app_name=f"{bot_name} ({platform.title()} Bot)",
            custom_instructions=platform_instructions.get(platform, "You are integrated into a chatbot platform.")
        )
    
    def respond_to_message(self, message: str, user_id: str = None) -> str:
        """
        Generate bot response to user message
        """
        context = f"Platform: {self.platform}"
        if user_id:
            context += f", User ID: {user_id}"
        
        result = self.fred.chat(
            message=message,
            context=context,
            conversation_id=user_id,  # Use user_id as conversation_id for user-specific history
            provider="anthropic"
        )
        
        return result.get("response", "Sorry, I couldn't process that request.")

# Example usage
if __name__ == "__main__":
    # Simple chat example
    print("🤖 Fix It Fred SDK Example")
    print("=" * 40)
    
    # Initialize simple chat
    fred_chat = FixItFredChat(provider="anthropic")
    
    # Test basic functionality
    response = fred_chat.ask("Hello! Can you help me with Python programming?")
    print(f"Fix It Fred: {response}")
    
    # Test SDK with different provider
    fred_sdk = FixItFredSDK()
    
    result = fred_sdk.chat(
        message="What's the weather like for coding today?",
        provider="openai",
        context="This is a test of the Fix It Fred SDK"
    )
    
    if result.get("success"):
        print(f"\nOpenAI Response: {result['response']}")
        print(f"Provider: {result['provider']}")
        print(f"Cached: {result.get('cached', False)}")
    
    # Check health
    health = fred_sdk.health_check()
    print(f"\nService Status: {health.get('status')}")
    print(f"Version: {health.get('version')}")